auth/auth-client/src/main/java/com/bazaarvoice/emodb/auth/proxy/CachingAuthenticatingProxy.java

```diff
@@ -16,33 +16,47 @@
 
     private final static int DEFAULT_CACHE_SIZE = 10;
 
-    private final LoadingCache<C, U> _cachedInstances;
+    private final Map<C, ProxyEntry> _cachedInstances;
+    private final int _cacheSize;
 
     public CachingAuthenticatingProxy() {
         this(DEFAULT_CACHE_SIZE);
     }
 
     public CachingAuthenticatingProxy(int cacheSize) {
-        checkArgument(cacheSize > 0, "Cache size must be positive");
-
-        _cachedInstances = CacheBuilder.newBuilder()
-                .maximumSize(cacheSize)
-                .build(new CacheLoader<C, U>() {
-                    @Override
-                    public U load(C credentials) throws Exception {
-                        return createInstanceWithCredentials(credentials);
-                    }
-                });
+        if (cacheSize <= 0) {
+            throw new IllegalArgumentException("Cache size must be positive");
+        }
+        _cachedInstances = new HashMap<>();
+        _cacheSize = cacheSize;
     }
 
     @Override
     public U usingCredentials(C credentials) {
         if (credentials == null) {
             throw new InvalidCredentialException("Credentials cannot be null");
         }
-        // Validate the credentials here and not in createInstanceWithCredentials(), otherwise the exception
-        // will be wrapped by the LoadingCache.
-        return _cachedInstances.getUnchecked(validateCredentials(credentials));
+        // Validate the credentials here and not in createInstanceWithCredentials() to ensure validation occurs
+        C validatedCredentials = validateCredentials(credentials);
+
+        ProxyEntry proxyEntry = _cachedInstances.get(validatedCredentials);
```

Comment:
This is all good, but honestly it's so rare that the same instance is used with more than one API key that it's probably fine to just keep the single MRU proxy rather than an LRU cache of more than 1.